Allow caution payments for invoices

jniles opened the pull request:
This PR presents the initial step in supporting cautions/prepayments in BHIMA.  The following components have been implemented:

 1. A reliable calculation of the patient's balance with the enterprise.  If a patient has a creditor balance with the hospital, the Patient Invoice module will signal the user that the patient has a creditor balance with an alert.
 2. Automatic "linking" of the patient's previous cash prepayment/caution when they are invoiced.  This is done by creating a third transaction that both debits and credits the same accountant.  Interestingly, this rule is not caught by the Trial Balance, which it probably should.

This PR is now complete!  It includes the option of toggling on/off prepayments, which will be particularly useful to institutions that do not accept prepayments.  The following issues have been closed:

 1. Closes #808 
 2. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/1324
 3. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/2687
 4. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/2695

jniles committed:
chore(patients): lint important client modules

This commit simply updates the patient client modules to pass eslint.

jniles committed:
feat(patients): add debtor balance endpoint

This commit adds an endpoint for the debtor balance route. Both the
route itself and the function existed previously, but now they are
hooked up.

jniles committed:
feat(invoices): add alert for creditor balance

This commit adds the client-side code to handle creditor balances.  It
moves the balance check into the setPatient() call, instead of leaving
it in the controller.

Eventually, this alert should only be shown in the enterprise settings
have the "autopay invoices using positive balance" set.

jniles committed:
feat(invoices): add prepayment allocation

This commit adds the SQL framework for linking prepayment after the
invoice is created by writing a third transaction that credits the
invoice and debits the caution, all hitting the same debtor and account.

TODO:
 - Translations on client and voucher description
 - Assign a "Caution Link" transaction type
 - Further testing

jniles committed:
fix: unit tests for PatientInvoiceForm

This commit updates paths associated with the PatientInvoiceForm to make
sure the $httpBackend captures the balance lookup path.

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -17,8 +16,7 @@ CREATE PROCEDURE StageInvoice(
   IN description TEXT,
-  IN service_id SMALLINT(5) UNSIGNED,
-  IN debtor_uuid BINARY(16),
+  IN service_id SMALLINT(5) UNSIGNED, IN debtor_uuid BINARY(16),
```
Why this in one line instead of multiple lines.

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -580,5 +578,196 @@ BEGIN
+    SELECT payment.uuid, payment.balance FROM stage_payment_balances AS payment;
+  DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;
+
+  CALL SetupDebugLog();
```
Is this call necessary ?

jniles replied:
Hehehe nope.  Good catch.

mbayopanda commented on sh/build-database.sh:
```diff
@@ -29,8 +29,9 @@ echo "[build] functions"
-mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/procedures.sql &> /dev/null
+mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/procedures.sql
 mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/admin.sql &> /dev/null
+mysql -u $DB_USER -p$DB_PASS $DB_NAME < server/models/debug.sql &> /dev/null
```
`debug.sql` is now a part of development environment.

mbayopanda commented on server/models/procedures/invoicing.sql:
```diff
@@ -580,5 +578,196 @@ BEGIN
+  CALL SetupDebugLog();
+
+  -- TODO(@jniles) - decide on this for real
+  SET linkTransactionTypeId = 3;
```
`TransactionTypeId = 3` is for convention payment, so maybe we can add a new transaction type for Caution

![caution link transaction type](https://user-images.githubusercontent.com/5445251/39124536-b45f3934-46f3-11e8-9997-748e2d7ceb20.png)

mbayopanda reviewed (commented):
I made a comment on `transactionType=3` apart from that the caution payment and its link to invoice work well.

jniles committed:
chore(enteprise): add setting for prepayments

This commit adds the prepayments setting to the database.

jniles committed:
feat(enterprise): toggle prepayments globally

This commit adds the ability to toggle on and off prepayments globally
throughout the application.  If you set "enable_prepayments" to true,
the following happens:
 1. The patient invoice will show alerts for creditor balances.
 2. The patient invoice module will automatically allocate the positive
 balance of the debtor's account to pay the invoice.
 3. The cash payments page will allow you to select a "caution" payment
 to prepay a future invoice.

If you set "enable_prepayments" to false:
 1. The patient invoice will never show an alert, regardless of if the
 debtor has a creditor balance.
 2. No automatic allocation of creditor balances occurs upon invoicing.
 3. There is no option to pay a caution payment on the Cash Window page.

Closes #2695.

mbayopanda commented on client/src/modules/cash/cash-form.service.js:
```diff
@@ -29,10 +29,16 @@ function CashFormService(AppCache, Session, Patients, Exchange) {
   CashForm.prototype.setup = function setup() {
     this.details = { date : new Date(), description : '', invoices : [] };
     this.details.currency_id = Session.enterprise.currency_id;
+    this.hasPrepaymentSupport = Session.enterprise.settings.enable_prepayments;
```
👍

mbayopanda reviewed (approved):
LGTM

Here are some things that i noticed : 
- In the enterprise module we can have a button for updating enterprise information in the section "Parametre de l'entreprise" for an easy use

- Translate are missing in the message of caution in the invoicing page.
